lib/web_ui/lib/src/engine/pointer_converter.dart

```diff
@@ -0,0 +1,255 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of engine;
+
+class _PointerState {
+  _PointerState(this.x, this.y);
+
+  int get pointer => _pointer; // The identifier used in framework hit test.
+  int _pointer;
+  static int _pointerCount = 0;
+  void startNewPointer() {
+    _pointerCount += 1;
+    _pointer = _pointerCount;
+  }
+
+  bool down = false;
+
+  double x;
+  double y;
+}
+
+/// Converter to convert web pointer data into a form that framework can
+/// understand.
+///
+/// This converter calculates pointer location delta and pointer identifier for
+/// each pointer. Both are required by framework to correctly trigger gesture
+/// activity. It also attempts to sanitize pointer data input sequence by always
+/// synthesizing an add pointer data prior to hover or down if it the pointer is
+/// not previously added.
+///
+/// For example:
+///   before:
+///     hover -> down -> move -> up
+///   after:
+///     add(synthesize) -> hover -> down -> move -> up
+///
+///   before:
+///     down -> move -> up
+///   after:
+///     add(synthesize) -> down -> move -> up
+class PointerDataConverter {
+  PointerDataConverter();
+
+  // Map from browser pointer identifiers to PointerEvent pointer identifiers.
+  final Map<int, _PointerState> _pointers = <int, _PointerState>{};
+
+  /// Clears the existing pointer states.
+  ///
+  /// This method is invoked during hot reload to make sure we have a clean
+  /// converter after hot reload.
+  void clearPointerState() {
+    _pointers.clear();
+    _PointerState._pointerCount = 0;
+  }
+
+  _PointerState _ensureStateForPointer(ui.PointerData datum, double x, double y) {
+    return _pointers.putIfAbsent(
+      datum.device,
+        () => _PointerState(x, y),
+    );
+  }
+
+  ui.PointerData _generateCompletePointerData(ui.PointerData datum) {
+    assert(_pointers.containsKey(datum.device));
+    final _PointerState state = _pointers[datum.device];
+    final double deltaX = datum.physicalX - state.x;
+    final double deltaY = datum.physicalY - state.y;
+    state.x = datum.physicalX;
+    state.y = datum.physicalY;
+    return ui.PointerData(
+      timeStamp: datum.timeStamp,
+      change: datum.change,
+      kind: datum.kind,
+      signalKind: datum.signalKind,
+      device: datum.device,
+      pointerIdentifier: state.pointer ?? 0,
```

Comment:
is it possible for `state.pointer` to be null?

Reply:
the state.pointer will only be assigned when the pointer has been down before. pointerIdentifier is only needed in down, move, up, and cancel. In all those cases, the pointer is guaranteed to be down. In other cases, pointerIdentifier will not be used.

Reply:
Nevermind. For some reason I confused the `_PointerState` constructor with `startNewPointer()` which always sets `_pointer` to some value.